src/processor/operator/copy/copy_node.cpp

```diff
@@ -81,145 +69,163 @@ std::pair<std::string, common::row_idx_t> CopyNode::getFilePathAndRowIdxInFile()
-    InMemColumnChunk* chunk, offset_t startOffset, uint64_t numValues) {
-    for (auto i = 0u; i < numValues; i++) {
-        auto offset = i + startOffset;
-        auto value = chunk->getValue<int64_t>(i);
-        if (!sharedState->pkIndex->append(value, offset)) {
-            return i;
+void CopyNode::executeInternal(ExecutionContext* context) {
+    // CopyNode goes through UNDO log, should be logged and flushed to WAL before making changes.
+    sharedState->logCopyNodeWALRecord(copyNodeInfo.wal);
+    while (children[0]->getNextTuple(context)) {
+        auto dataChunkToCopy = resultSet->getDataChunk(0);
+        // All tuples in the resultSet are in the same data chunk.
+        auto numTuplesToAppend = ArrowColumnVector::getArrowColumn(
+            resultSet->getValueVector(copyNodeInfo.dataColumnPoses[0]).get())
+                                     ->length();
+        uint64_t numAppendedTuples = 0;
+        while (numAppendedTuples < numTuplesToAppend) {
+            numAppendedTuples += localNodeGroup->append(
+                resultSet, copyNodeInfo.dataColumnPoses, numTuplesToAppend - numAppendedTuples);
+            if (localNodeGroup->getNumNodes() == StorageConstants::NODE_GROUP_SIZE) {
+                // Current node group is full, flush it and reset it to empty.
+                auto nodeGroupIdx = sharedState->getNextNodeGroupIdx();
+                localNodeGroup->setNodeGroupIdx(nodeGroupIdx);
+                appendNodeGroupToTableAndPopulateIndex(sharedState->table, localNodeGroup.get(),
+                    sharedState->pkIndex.get(), sharedState->pkColumnID);
+            }
+            if (numAppendedTuples < numTuplesToAppend) {
+                auto slicedChunk = sliceDataVectorsInDataChunk(*dataChunkToCopy,
+                    copyNodeInfo.dataColumnPoses, (int64_t)numAppendedTuples,
+                    (int64_t)(numTuplesToAppend - numAppendedTuples));
+                resultSet->dataChunks[0] = slicedChunk;
+            }
         }
     }
-    return numValues;
+    // Append left data in the local node group to the shared one.
+    if (localNodeGroup->getNumNodes() > 0) {
+        sharedState->appendLocalNodeGroup(std::move(localNodeGroup));
+    }
 }
 
-template<>
-uint64_t CopyNode::appendToPKIndex<ku_string_t, InMemOverflowFile*>(InMemColumnChunk* chunk,
-    offset_t startOffset, uint64_t numValues, InMemOverflowFile* overflowFile) {
-    for (auto i = 0u; i < numValues; i++) {
-        auto offset = i + startOffset;
-        auto value = chunk->getValue<ku_string_t>(i);
-        auto key = overflowFile->readString(&value);
-        if (!sharedState->pkIndex->append(key.c_str(), offset)) {
-            return i;
-        }
-    }
-    return numValues;
+std::shared_ptr<DataChunk> CopyNode::sliceDataVectorsInDataChunk(const DataChunk& dataChunkToSlice,
+    const std::vector<DataPos>& dataColumnPoses, int64_t offset, int64_t length) {
+    auto slicedChunk =
```

Comment:
The slicing algorithm looks a little inefficient since we need to create a new dataChunk and vectors for each slice (even though the arrowColumn::slice is zero-copy).

Reply:
```
/// Construct a zero-copy slice of the array with the indicated offset and
  /// length
  ///
  /// \param[in] offset the position of the first element in the constructed
  /// slice
  /// \param[in] length the length of the slice. If there are not enough
  /// elements in the array, the length will be adjusted accordingly
  ///
  /// \return a new object wrapped in std::shared_ptr<Array>
  std::shared_ptr<Array> Slice(int64_t offset, int64_t length) const;
```
The documentation of arrowColumn::slice states that it is zero-copy